### PR TITLE
[EMCAL-610] Output raw stream based on RawFileWriter

### PR DESCRIPTION
**Pull request by mfasDa**
- Replace usage of RawOutputPageHandler with
  RawFileWriter for output streaming
- Initialize RawFileWriter using a single output file and a
  linear link assignment for C-RORC/Link ID untill final
  link assignment on production FLPs is fixed

### CODE REVIEW

**Comment by mfasDa** [comment]
Pull request will be extended by a stand-alone application converting digit files to raw data.

**Comment by mfasDa** [comment]
@shahor02 @sawenzel The first IR should be the same for the digitizer and for the raw writer, how do we propagate it in the best way.

Currently for EMCAL the digits vector contains all digits in the timeframe, separated by triggers using a second vector with a range reference marking the digits belonging to the same readout window (trigger). This is done in order to implement the handling of pileup in the EMCAL. Live/Dead times are handled there as well. In the digitizer we use the collision context to get the IR, set the detector live for the first trigger, accumulate digits from collisions within the 1.5 ms readout window digitised using the EMCAL time response, and set the detector busy afterwards. Therefore even though for the moment the first collision opens the EMCAL readout window we cannot blindly use the first IR in the timeframe as in the future it might not be the first collision.

**Comment by shahor02** [comment]
Hi @mfasDa 

I have not really understood the question. The 1st IR of digitization can be anything, it is where the collision happened (currently the sampling starts for orbit/bc = 1/0, can be overridden by  --firstOrbit, --firstBC options to digitizer), the 1st IR of raw data is the start of the 1st TF, i.e. detectors data packaging (defined by the configurable param https://github.com/AliceO2Group/AliceO2/blob/b5b538995078db577663801c0bce9b55be64410b/Detectors/Raw/include/DetectorsRaw/HBFUtils.h#L126-L128). For consistence, the 1st TF 1st IR must be <= first collision IR.

Cheers,
 Ruben

**Comment by mfasDa** [comment]
@shahor02 The problem is that our readout window with 1.5 mus >> bunch spacing with 25 ns. Therefore in the digitization in order to simulate pileup we integrate over as many collisions as we can fit into the readout window (-600 ns due to delays in the cable). Therefore we decided to store the digits not per event - which would be anyhow questionable with pileup - but per timeframe, splitting the digits vector in ranges belonging to the same readout window. Therefore we assume in the digitization that all collisions we loop over belong to the same timeframe, and we push the digits vector once per timeframe. Doing this, we must have the same timing definition in the digitizer and the raw writer. We could simply mark the first IR but it might be that EMCAL will not trigger at the first IR and therefore set a wrong IR as first IR - which should be common for all detectors. Otherwise I would be curious how other detectors do it - in particular the continuous detectors must integrate in the digitization over many events and are busy at the end of timeframe, no?

**Comment by shahor02** [comment]
@mfasDa: all detectors store their data in this way, ITS readout window can be as large as 20 \mus, it does not matter. On the digitization level we don't have TF definition at all, but every digit is associated with well defined BC (when the trigger or strobe started). It is irrelevant what other (later) BCs will contribute to this digit, at digit->raw conversion it goes to the TF which its opening BC belongs to.

**Comment by mfasDa** [comment]
@shahor02 OK, for us all the digits in the same open readout window have the same IR, the one of the trigger which opened the readout window, and the raw writer makes use of it. Shall we rather push digits in readout windows?

**Comment by shahor02** [comment]
Is not the digit already covering exactly a readout window? What do you mean pushing them in readout window? In terms of the RawWriter you add a payload for the IR of the beginning of the readout window.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

File: Detectors/EMCAL/simulation/src/RawCreator.cxx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  bpo::positional_options_description opt_pos;
+
+  try {
+    auto add_option = opt_general.add_options();
+    add_option("help,h", "Print this help message");
+    add_option("verbose,v", bpo::value<uint32_t>()->default_value(0), "Select verbosity level [0 = no output]");
+    add_option("input-file,i", bpo::value<std::string>()->required(), "Specifies digit input file.");
+    add_option("output-file,o", bpo::value<std::string>()->required(), "Specifies raw ouput file.");
+    add_option("debug,d", bpo::value<uint32_t>()->default_value(0), "Select debug output level [0 = no debug output]");
+
```
Comment (shahor02) [inline Detectors/EMCAL/simulation/src/RawCreator.cxx:46]: Could you please add the following options:
1) possibility to produce 1 raw file per link (i.e. FEEID), this is preferred mode for replay using readout.exe. See how it is done e.g. for ITS:
https://github.com/AliceO2Group/AliceO2/blob/b0fc846fae2e7f3b1529ebacdb07c94347331b47/Detectors/ITSMFT/ITS/simulation/src/digi2raw.cxx#L61

2) parsing ``configKeyValues`` options, i.e.
https://github.com/AliceO2Group/AliceO2/blob/b0fc846fae2e7f3b1529ebacdb07c94347331b47/Detectors/ITSMFT/ITS/simulation/src/digi2raw.cxx#L65
+
https://github.com/AliceO2Group/AliceO2/blob/b0fc846fae2e7f3b1529ebacdb07c94347331b47/Detectors/ITSMFT/ITS/simulation/src/digi2raw.cxx#L85
This is necessary for e.g. generation of raw data with custom number of HBFs per TF.

Also, all detectors generate in the output directory a configuration file which can be provided to RawFileReader, this can be done by a simple call: https://github.com/AliceO2Group/AliceO2/blob/b0fc846fae2e7f3b1529ebacdb07c94347331b47/Detectors/ITSMFT/ITS/simulation/src/digi2raw.cxx#L173

Comment (mfasDa) [reply Detectors/EMCAL/simulation/src/RawCreator.cxx]: All requests are added to the RawCreator + RawWriter is adapted.

File: Detectors/EMCAL/simulation/src/RawWriter.cxx [hunk 2]
```diff
@@ -244,4 +245,29 @@ std::vector<char> RawWriter::createRCUTrailer(int payloadsize, int feca, int fec
   std::vector<char> encoded(trailerwords.size() * sizeof(uint32_t));
   memcpy(encoded.data(), trailerwords.data(), trailerwords.size() * sizeof(uint32_t));
   return encoded;
+}
+
+
+int RawWriter::carryOverMethod(const header::RDHAny* rdh, const gsl::span<char> data,
+                                            const char* ptr, int maxSize, int splitID,
+                                            std::vector<char>& trailer, std::vector<char>& header) const {
+  int offs = ptr - &data[0]; // offset wrt the head of the payload
```
Comment (shahor02) [inline Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: @mfasDa concerning the ``carryOverMethod``: does the original data you provide to ``writer->addData(..)`` already contain the trailer (placeholder) which need to be finalized in  ``carryOverMethod`` or you want to build it on the fly add and to every page payload, regardless if it the data need splitting or not? 
What we discussed yesterday is to call ``carryOverMethod`` also for the last chunk if there was at least 1 splitting, but given this will be wrong if your data comes w/o trailer.
In the latter case, I would simply add an option (you will need to activate it for EMCal) to always call this method, even if splitting is not needed at all.
But if your original data contain a trailer, then it is more tricky, since it will need to be modified in the original data which is supposed to be ``const``...

Additional point: the ``trailer`` string supplied by the method must be a multiple of 16B CRU word.
Is this is a case for the RCU trailer your create?
Also, if your original data do not contain a trailer (so it is added on top of the maxSize): 
can the trailer size exceed one 16B word? In this case, if the ``maxSize`` estimated by the writer is 16B (minimal possible value), then you will get a negative ``actualSize``.

Comment (mfasDa) [reply Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: @shahor02 Currently the trailer is added at the end of the payload, containing the full payload size which is the number of 32bit ALTRO words in the payload sent to addData. In the hardware the trailer (size typically 9 32 bit ADC words) is added to every page and it contains the number of 10 bit words per page, so in the decoding I have to add the payload size from all the trailers belonging to the same page in order  to know how the exact amount of words. In the simulation the trailers must be handled in the same way.

In the carryOverMethod implementation used now I read the trailer from the end-of-payload and replace the payload size with the payload size of the page, in the same way as it is done for ITSMFT. (This answers the first question with yes) This is rather trivial as the number of 8bit words just needs to be divided by 4. However the last trailer must be adapted as well and this I can only do once I know the size of the payload in the last page.

Concerning maxSize: I was under the impression this is the size of the page? The trailer would always be smaller. Did I misunderstand something?

Comment (shahor02) [reply Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: Hi @mfasDa
The maxSize is the number of bytes remaining in the current open page to reach 8KB limit (or the super-page to reach is limit, by default 1MB). So, ultimately, the maxSize can be just 16B. But if your data already contains the fixed-size trailer, this is not a problem. Since you are starting a new page for every trigger, I don't think you will have a situation of only 16B remaining on the 8KB page, it could happen if you add a new payload to a partially filled page. And in case of super-page filling there is a protection to not create a new page in the same super-page if less than 128B left (from which 64 will be taken by the RDH), so, you should be safe.

Basically, for the EMCAL I will need the writer will need to call carryOverMethod for the very last chunk (where no splitting is actually needed) and instead of adding the supplied trailer after writing the payload (as it would do if there is a real splitting) it should overwrite the last ``carryOverTrailer.size()`` bytes, is this correct?

Comment (shahor02) [reply Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: @mfasDa please check if https://github.com/AliceO2Group/AliceO2/pull/4179 is what you want. You will need to enable the carry-over call for the last chunk by ``RawFileWriter::setApplyCarryOverToLastPage(true);`` and detect such a call as the modified test shows: https://github.com/AliceO2Group/AliceO2/blob/863c3435580f5fbe1bc34870f4a40a31f45aea82/Detectors/Raw/test/testRawReaderWriter.cxx#L166-L170

Comment (mfasDa) [reply Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: Hi @shahor02, yes, the conclusion is correct. The carryOverMethod currently sets the payload size to the trailer whenever it is called. So since the end of data contains the trailer either it has to be overwritten by the RawFileWriter or our custom carryOverMethod. In the second case I would not add a trailer but modify it myself (probably overall easier). In this case I would need to know that we are on the last page.

Comment (shahor02) [reply Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: @mfasDa as far as I can see, your carryOverMethod sets the returned ``actual`` size to ``maxSize - trailer.size``, this is ok except when you are on the last chunk: in this case the you should return the original ``maxSize``  as an actual payload to write and supply the modified trailer string. The writer will overtwrite last  ``trailer.size()`` of the payload last chunk by this modified trailer. You cannot modify in the carry-over method the original data: it is ``const gsl::span``, just provide an updated trailer to override the one in the payload.

Comment (mfasDa) [reply Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: OK, so I only need to know that I am on the last page to include the trailer size in the from the actual size. I was thinking to take this from the payload size but in principle the page could be 8 kB. Anything that tells me how much data is still to come? I.e. the gsl::span contains the full data or the data starting from the page?

Comment (shahor02) [reply Detectors/EMCAL/simulation/src/RawWriter.cxx:272]: The ``const gsl::span<char> data`` refers to the full original payload you provided to ``addData``, ``const char* ptr``
is start of data yet to write (can be anything between ``data.begin() and data.end()``) and the ``int maxSize`` is the max number of bytes the writer can write (starting from the ``ptr``) to currently open 8kb page.
You can detect that you are on the last page by https://github.com/AliceO2Group/AliceO2/blob/72e4fba376d9cff8b1677b5cd93982123496d9ba/Detectors/Raw/test/testRawReaderWriter.cxx#L166-L167

**Review by shahor02 — CHANGES_REQUESTED** [review]
@mfasDa : you should also set (e.g. in the init) ``writer->setApplyCarryOverToLastPage(true)`` to trigger the callback at the last page.

**Comment by mfasDa** [comment]
@shahor02 I intended to wait until #4179 is merged and then add the line in order to avoid conflicts. No I see it is merged so I will add this immediately.
